### ISSUE TITLE
Can not rename Input_boolean name in UI

### ISSUE THREAD

**Issue by Mariusthvdb** [home-assistant/frontend#15906]
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

<img width="591" alt="Scherm­afbeelding 2023-03-22 om 22 08 15" src="https://user-images.githubusercontent.com/33354141/227039159-c419de6f-2a76-4316-80b2-1949a9fb09ff.png">

<img width="594" alt="Scherm­afbeelding 2023-03-22 om 22 09 29" src="https://user-images.githubusercontent.com/33354141/227039210-9e06f551-c419-4253-8cb4-60bc9e4600d5.png">

<img width="587" alt="Scherm­afbeelding 2023-03-22 om 22 09 40" src="https://user-images.githubusercontent.com/33354141/227039266-be83e514-2401-4427-b7dc-2565b03dda93.png">

### Describe the behavior you expected

successful name change 

### Steps to reproduce the issue

1. Create boolean
2. change name
3. tikte remains only name
...

### What version of Home Assistant Core has the issue?

2023.3.5

### In which browser are you experiencing the issue with?

any

### Which operating system are you using to run this browser?

Ventura 13.2

**Comment by piitaya** [comment]
Not sure if it's a core issue or a frontend issue. 
To reproduce it : 
- create a input boolean
- change the entity id
- save
- rename it
- save
- the name is not updated in the registry entry.

**Comment by Mariusthvdb** [comment]
Yeah, guess that's up to you. I merely reported it here because experienced in the Ui.

Not sure if you  expect anything further now?

**Comment by karwosts** [comment]
I poked at this for a little bit and I'm pretty sure its a core issue. 

I chased this way deep into the weeds of the helper collections, and I haven't quite found the issue yet but the thing that looks off to me is this code:

```
@callback
def sync_entity_lifecycle(
    hass: HomeAssistant,
    domain: str,
    platform: str,
    entity_component: EntityComponent,
    collection: StorageCollection | YamlCollection,
    entity_class: type[CollectionEntity],
) -> None:
    """Map a collection to an entity component."""
    entities: dict[str, CollectionEntity] = {}
    ent_reg = entity_registry.async_get(hass)

    async def _add_entity(change_set: CollectionChangeSet) -> CollectionEntity:
        def entity_removed() -> None:
            """Remove entity from entities if it's removed or not added."""
            if change_set.item_id in entities:
                entities.pop(change_set.item_id)

        entities[change_set.item_id] = collection.create_entity(
            entity_class, change_set.item
        )
        entities[change_set.item_id].async_on_remove(entity_removed)
        return entities[change_set.item_id]

    async def _remove_entity(change_set: CollectionChangeSet) -> None:
        ent_to_remove = ent_reg.async_get_entity_id(
            domain, platform, change_set.item_id
        )
        if ent_to_remove is not None:
            ent_reg.async_remove(ent_to_remove)
        elif change_set.item_id in entities:
            await entities[change_set.item_id].async_remove(force_remove=True)
        # Unconditionally pop the entity from the entity list to avoid racing against
        # the entity registry event handled by Entity._async_registry_updated
        if change_set.item_id in entities:
            entities.pop(change_set.item_id)

    async def _update_entity(change_set: CollectionChangeSet) -> None:
        if change_set.item_id not in entities:
            return
        await entities[change_set.item_id].async_update_config(change_set.item)
```

The dict "entities" here typically has one entry for each entity in a domain. 

When an entity (e.g.) an input boolean has its entity id changed, it is removed from this dict but *it does not get re-added*.

Then on subsequent name changes, _update_entity is called but it takes the early return because item_id not in entities. And therefore async_update_config is not called. 


If I reboot HA, all the entities that are missing from the collection get repopulated, and then I can rename entities with non-original entity_ids again. 

So the inability to rename seems to be transient, only from the time when the entity_id is changed, to the time HA is rebooted, then it fixes itself. 

Need to understand why entity is being removed from this collection but not re-added.
